tarofreighter/chain_porter.go

```diff
@@ -491,79 +491,41 @@ func (p *ChainPorter) transferReceiverProof(pkg *sendPackage) error {
 }
 
 // advanceState advances the state machine.
-func (p *ChainPorter) advanceState(currentPkg *sendPackage,
+func (p *ChainPorter) advanceState(pkg *sendPackage,
 	txBroadcastRespChan chan *PendingParcel) error {
 
-	logState := func(sendState SendState) {
-		log.Infof("ChainPorter executing state: %v", sendState)
-	}
-
-	// Advance through sync evaluated states. (Up to, but not including,
-	// SendStateWaitTxConf.)
-	continueStateProgression := currentPkg.SendState < SendStateWaitTxConf
-	for continueStateProgression {
-		logState(currentPkg.SendState)
-
-		// Before we attempt a state transition, make sure that we
-		// aren't trying to shut down.
-		select {
-		case <-p.Quit:
-			return fmt.Errorf("porter shutting down")
-
-		default:
-		}
-
-		updatedPkg, err := p.stateStep(*currentPkg, txBroadcastRespChan)
-		if err != nil {
-			return err
-		}
-
-		// Continue state progression whilst target state has not yet
-		// been reached. Stop before evaluating target state.
-		continueStateProgression = updatedPkg.SendState < SendStateWaitTxConf
-
-		currentPkg = updatedPkg
-	}
-
-	// The remaining state transitions are evaluated using a goroutine
-	// because tx confirmation and proof transfer could take a while.
+	// Some states require a goroutine because tx confirmation and proof
```

Comment:
Nice, this is something I was going to comment on, that it would be great to have everything in a single state machine.
_But_, this changes the behavior from the point of view of the RPC client: Before, we'd only return from the send RPC call once the state `SendStateBroadcast` was reached. And any error occurring before that would be reported back synchronously to the RPC client (I think). Now, because we're entering asynchronous code directly, any error happening before broadcasting would only be visible in the log.

Reply:
As far as I understand, the situation as you've described is still true even in the latest commits. I suppose one way forward is for the client to subscribe to event notifications as a way of getting updates on the async progress.

But I don't think errors are currently communicated on that stream. I will make an issue to included errors in the event stream. Let me know if you have further thoughts on this.